File: apps/cargo-tracker/app/cargotracker/tracking/domain/model/valueobjects/TrackingBookingId.scala
package cargotracker.tracking.domain.model.valueobjects

/** Tracking Context 内で扱う予約 ID。Booking Context の `BookingId` とは別型（コンテキスト分離 / ADR 0010）。 */
opaque type TrackingBookingId = String

object TrackingBookingId:
  sealed trait Error
  case object Empty extends Error

  def apply(raw: String): Either[Error, TrackingBookingId] =
    if raw.nonEmpty then Right(raw) else Left(Empty)

  def unsafeFrom(raw: String): TrackingBookingId =
    apply(raw).fold(_ => throw IllegalArgumentException(s"Invalid TrackingBookingId: $raw"), identity)

  extension (id: TrackingBookingId) def value: String = id

File: apps/cargo-tracker/src/main/java/com/example/cargotracker/tracking/domain/model/TrackingBookingId.java
package com.example.cargotracker.tracking.domain.model;

import java.util.UUID;

/**
 * 予約参照 ID（Tracking Context 固有の型）。
 *
 * <p>Booking の {@code BookingId} を参照しない（ADR-005・ArchUnit ルール 4）。
 * 追跡が知る必要があるのは「どの予約の追跡か」という事実だけである。
 *
 * @param value 予約 ID
 */
public record TrackingBookingId(UUID value) {

    public TrackingBookingId {
        if (value == null) {
            throw new IllegalArgumentException("予約 ID は必須です");
        }
    }
}
